### PR TITLE
[go_router_builder] Removes `path_to_regexp` from the dependencies 

### PR DESCRIPTION
**Pull request by ValentinVignal**
Fixes https://github.com/flutter/flutter/issues/130817
Relates to https://github.com/flutter/flutter/issues/122713

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/go_router_builder/lib/src/path_utils.dart
Comment (ValentinVignal) [inline packages/go_router_builder/lib/src/path_utils.dart:1]: This is a literal copy past from `go_router`'s `path_utils.dart` file https://github.com/flutter/packages/blob/main/packages/go_router/lib/src/path_utils.dart

File: packages/go_router_builder/lib/src/path_utils.dart [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+///  `pattern` = `/user/:id/book/:bookId`
+///
+///  The `parameters` would contain `['id', 'bookId']` as a result of calling
+///  this method.
+///
+/// To extract the path parameter values from a [RegExpMatch], pass the
+/// [RegExpMatch] into [extractPathParameters] with the `parameters` that are
+/// used for generating the [RegExp].
+RegExp patternToRegExp(String pattern, List<String> parameters) {
+  final StringBuffer buffer = StringBuffer('^');
```
Comment (ValentinVignal) [inline packages/go_router_builder/lib/src/path_utils.dart]: This file is this commit ([chore: Copy path_utils.dart from go_router](https://github.com/flutter/packages/pull/4524/commits/c9487b301293bedc308e1b4561055db1f9ec6e61)) a literal copy past from `go_router`'s `path_utils.dart` file https://github.com/flutter/packages/blob/main/packages/go_router/lib/src/path_utils.dart

File: packages/go_router_builder/lib/src/path_utils.dart [hunk 2]
```diff
@@ -87,74 +85,3 @@ String patternToPath(String pattern, Map<String, String> pathParameters) {
...
-Map<String, String> extractPathParameters(
-    List<String> parameters, RegExpMatch match) {
-  return <String, String>{
-    for (int i = 0; i < parameters.length; ++i)
-      parameters[i]: match.namedGroup(parameters[i])!
-  };
-}
-
-/// Concatenates two paths.
-///
```
Comment (ValentinVignal) [inline packages/go_router_builder/lib/src/path_utils.dart]: In this commit ([revamp: Remove unused methods](https://github.com/flutter/packages/pull/4524/commits/0cc50db8b064e6b26629891a16710ae3272af622)) I'm only deleting methods that are not used

File: packages/go_router_builder/lib/src/path_utils.dart [hunk 3]
```diff
@@ -4,70 +4,29 @@
...
-
-  if (start < pattern.length) {
-    buffer.write(RegExp.escape(pattern.substring(start)));
-  }
-
-  if (!pattern.endsWith('/')) {
-    buffer.write(r'(?=/|$)');
-  }
-  return RegExp(buffer.toString(), caseSensitive: false);
-}
```
Comment (ValentinVignal) [inline packages/go_router_builder/lib/src/path_utils.dart]: In this commit ([refactor: Simplify path utils methods to only keep what is used](https://github.com/flutter/packages/pull/4524/commits/205da049bc308c5643ad6fdf62b19e26e8a0bff0)),

I'm only deleting and simplifying `patternToRegExp`. We don't need to used the returnd `RegExp`. So instead I'm returning the `parameters` (and I don't need the parameter `parameters` anymore)

**Comment by ValentinVignal** [comment]
I believe this PR can have a no-test exemption. This is a revamp and the current tests should cover it already

File: packages/go_router_builder/pubspec.yaml [hunk 4]
```diff
@@ -2,7 +2,7 @@ name: go_router_builder
 description: >-
   A builder that supports generated strongly-typed route helpers for
   package:go_router
-version: 2.2.2
+version: 2.3.0
```
Comment (chunhtai) [inline packages/go_router_builder/pubspec.yaml:5]: There isn't a new feature, should probably update patch version.

**Review by chunhtai — COMMENTED** [review]
can you remove the path_to_regexp in script/configs/allowed_unpinned_deps.yaml, too?

**Comment by ValentinVignal** [comment]
> can you remove the path_to_regexp in script/configs/allowed_unpinned_deps.yaml, too?

I removed it in [ci: Remove path_to_regexp from allowed unpinned deps](https://github.com/flutter/packages/pull/4524/commits/5969d6fbb8cb5807ec92ee23636356fc7f5739d5)

File: packages/go_router_builder/pubspec.yaml (hunk 4 above)
Comment (ValentinVignal) [reply packages/go_router_builder/pubspec.yaml]: Sure, I did it in [chore: Only update patch version](https://github.com/flutter/packages/pull/4524/commits/57f71a960370c0da9b138b0e58d8fe6bce0cdeed)

**Comment by chunhtai** [comment]
It looks like the branch is outdated so the ci.yaml failed. can you rebase off latest main?

**Review by chunhtai — APPROVED** [review]
LGTM

**Review by stuartmorgan — APPROVED** [review]
LGTM, thanks!

**Review by hangyujin — APPROVED** [review]
LGTM

**Comment by Hixie** [comment]
it's probably worth throwing in a few unit tests for the functions in packages/go_router_builder/lib/src/path_utils.dart, just so that the next person to refactor them can be confident they didn't break anything subtle about that code.

**Review by chunhtai — CHANGES_REQUESTED** [review]
As @Hixie stated, please add some tests and re-request my review once done

**Comment by ValentinVignal** [comment]
> tests: Add path_utils unit test

I added some unit tests in [tests: Add path_utils unit test](https://github.com/flutter/packages/pull/4524/commits/d8391de35d1915992d46414328b8401fd226a112)

**Review by chunhtai — APPROVED** [review]
LGTM

**Pull request merged**